### PR TITLE
Remove own impls of Readonlylist and dict

### PR DESCRIPTION
**Pull request by teinarss**

### CODE REVIEW

File: OpenRA.Game/GameRules/Ruleset.cs [hunk 1]
```diff
@@ -117,7 +117,7 @@ public class Ruleset
 			if (filterNode != null)
 				yamlNodes = yamlNodes.Where(k => !filterNode(k));
 
-			return new ReadOnlyDictionary<string, T>(yamlNodes.ToDictionaryWithConflictLog(k => k.Key.ToLowerInvariant(), makeObject, "LoadFromManifest<" + name + ">"));
+			return new Dictionary<string, T>(yamlNodes.ToDictionaryWithConflictLog(k => k.Key.ToLowerInvariant(), makeObject, "LoadFromManifest<" + name + ">"));
```
Comment (pchote) [inline OpenRA.Game/GameRules/Ruleset.cs]: Isn't `return new yamlNodes.ToDictionaryWithConflictLog(...);` sufficient here?

File: OpenRA.Game/Graphics/HardwarePalette.cs [hunk 2]
```diff
@@ -28,7 +28,7 @@ public sealed class HardwarePalette : IDisposable
 		public HardwarePalette()
 		{
 			Texture = Game.Renderer.Context.CreateTexture();
-			readOnlyModifiablePalettes = modifiablePalettes.AsReadOnly();
+			readOnlyModifiablePalettes = modifiablePalettes;
```
Comment (pchote) [inline OpenRA.Game/Graphics/HardwarePalette.cs]: Looks like `readOnlyModifiablePalettes` can be removed completely now, just pass `modifiablePalettes` directly in the one place this is used. Maybe also rename `modifiablePalettes` to `mutablePalettes` in a separate commit?

File: OpenRA.Game/Manifest.cs [hunk 3]
```diff
@@ -213,10 +214,10 @@ static string[] YamlList(Dictionary<string, MiniYaml> yaml, string key, bool par
 		static IReadOnlyDictionary<string, string> YamlDictionary(Dictionary<string, MiniYaml> yaml, string key)
 		{
 			if (!yaml.ContainsKey(key))
-				return new ReadOnlyDictionary<string, string>();
+				return new Dictionary<string, string>();
 
 			var inner = yaml[key].ToDictionary(my => my.Value);
-			return new ReadOnlyDictionary<string, string>(inner);
+			return inner;
```
Comment (pchote) [inline OpenRA.Game/Manifest.cs]: `return yaml[key].ToDictionary(my => my.Value);`

File: OpenRA.Mods.Cnc/Traits/TDGunboat.cs [hunk 4]
```diff
@@ -42,7 +43,7 @@ IEnumerable<ActorInit> IActorPreviewInitInfo.ActorPreviewInits(ActorInfo ai, Act
 		public IReadOnlyDictionary<CPos, SubCell> OccupiedCells(ActorInfo info, CPos location, SubCell subCell = SubCell.Any)
 		{
 			var occupied = new Dictionary<CPos, SubCell>() { { location, SubCell.FullCell } };
-			return new ReadOnlyDictionary<CPos, SubCell>(occupied);
+			return occupied;
```
Comment (pchote) [inline OpenRA.Mods.Cnc/Traits/TDGunboat.cs]: `return new Dictionary<CPos, SubCell>() { { location, SubCell.FullCell } };`

File: OpenRA.Mods.Common/Traits/Buildings/Building.cs [hunk 5]
```diff
@@ -241,7 +242,7 @@ public virtual bool IsCloseEnoughToBase(World world, Player p, ActorInfo ai, CPo
 			var occupied = OccupiedTiles(topLeft)
 				.ToDictionary(c => c, c => SubCell.FullCell);
 
-			return new ReadOnlyDictionary<CPos, SubCell>(occupied);
+			return occupied;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Buildings/Building.cs]: `return` the line above directly here too.

File: OpenRA.Mods.Common/Traits/Crates/Crate.cs [hunk 6]
```diff
@@ -35,7 +35,7 @@ public class CrateInfo : TraitInfo, IPositionableInfo, Requires<RenderSpritesInf
 		public IReadOnlyDictionary<CPos, SubCell> OccupiedCells(ActorInfo info, CPos location, SubCell subCell = SubCell.Any)
 		{
 			var occupied = new Dictionary<CPos, SubCell>() { { location, SubCell.FullCell } };
-			return new ReadOnlyDictionary<CPos, SubCell>(occupied);
+			return occupied;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Crates/Crate.cs]: and here.

File: OpenRA.Mods.Common/Traits/Husk.cs [hunk 7]
```diff
@@ -37,7 +38,7 @@ IEnumerable<ActorInit> IActorPreviewInitInfo.ActorPreviewInits(ActorInfo ai, Act
 		public IReadOnlyDictionary<CPos, SubCell> OccupiedCells(ActorInfo info, CPos location, SubCell subCell = SubCell.Any)
 		{
 			var occupied = new Dictionary<CPos, SubCell>() { { location, SubCell.FullCell } };
-			return new ReadOnlyDictionary<CPos, SubCell>(occupied);
+			return occupied;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Husk.cs]: and here.

File: OpenRA.Mods.Common/Traits/Immobile.cs [hunk 8]
```diff
@@ -25,7 +24,7 @@ class ImmobileInfo : TraitInfo, IOccupySpaceInfo
 			var occupied = OccupiesSpace ? new Dictionary<CPos, SubCell>() { { location, SubCell.FullCell } } :
 				new Dictionary<CPos, SubCell>();
 
-			return new ReadOnlyDictionary<CPos, SubCell>(occupied);
+			return occupied;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Immobile.cs]: here too.

File: OpenRA.Mods.Common/Traits/World/EditorActorPreview.cs [hunk 9]
```diff
@@ -80,8 +81,7 @@ public EditorActorPreview(WorldRenderer worldRenderer, string id, ActorReference
 				Footprint = ios.OccupiedCells(Info, location, subCell);
 			else
 			{
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/World/EditorActorPreview.cs]: Can remove these braces now, or possibly replace this whole thing with
```csharp
Footprint = ios?.OccupiedCells(Info, location, subCell) ?? new Dictionary<CPos, SubCell>() { { location, SubCell.FullCell } };
```
(but maybe a little long?)

File: OpenRA.Mods.Common/Traits/World/MapOptions.cs [hunk 10]
```diff
@@ -82,14 +82,14 @@ IEnumerable<LobbyOption> ILobbyOptions.LobbyOptions(Ruleset rules)
 
 			if (techLevels.Any())
 				yield return new LobbyOption("techlevel", TechLevelDropdownLabel, TechLevelDropdownDescription,	TechLevelDropdownVisible, TechLevelDropdownDisplayOrder,
-					new ReadOnlyDictionary<string, string>(techLevels),	TechLevel, TechLevelDropdownLocked);
+					techLevels,	TechLevel, TechLevelDropdownLocked);
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/World/MapOptions.cs]: Can we please fix this tab &rarr; space while we're here?

File: OpenRA.Mods.Cnc/SpriteLoaders/ShpTDLoader.cs [hunk 11]
```diff
@@ -156,6 +155,8 @@ public ShpTDSprite(Stream stream)
 
 			foreach (var h in headers)
 				Decompress(h);
+
+			Frames = headers.ToList();
```
Comment (pchote) [inline OpenRA.Mods.Cnc/SpriteLoaders/ShpTDLoader.cs]: Is the `ToList` actually needed here?

File: OpenRA.Mods.Common/SpriteLoaders/DdsLoader.cs [hunk 12]
```diff
@@ -74,7 +75,7 @@ public DdsFrame(Stream stream)
 
 		public DdsSprite(Stream stream)
 		{
-			Frames = new ISpriteFrame[] { new DdsFrame(stream) }.AsReadOnly();
+			Frames = new ISpriteFrame[] { new DdsFrame(stream) }.ToList();
```
Comment (pchote) [inline OpenRA.Mods.Common/SpriteLoaders/DdsLoader.cs]: I'm more confident that the `ToList` isn't needed here.

File: OpenRA.Mods.Common/SpriteLoaders/TgaLoader.cs [hunk 13]
```diff
@@ -108,7 +109,7 @@ public TgaFrame(Stream stream, Size frameSize, Rectangle frameWindow)
 
 		public TgaSprite(Stream stream)
 		{
-			Frames = new ISpriteFrame[] { new TgaFrame(stream) }.AsReadOnly();
+			Frames = new ISpriteFrame[] { new TgaFrame(stream) }.ToList();
```
Comment (pchote) [inline OpenRA.Mods.Common/SpriteLoaders/TgaLoader.cs]: Likewise here.

**Review by pchote — COMMENTED** [review]
Feels good to remove long-dead code!

Here are some style nits that the automated refactoring wasn't smart enough to deal with:

**Comment by teinarss** [comment]
Updated.

**Review by pchote — APPROVED** [review]

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/2dc6955951a695f832966d99d054a6d26f0361a6...4ea03ab06c0ffc6a68928c0cabad52fdec5413fd

**Pull request merged**